File: MyXargs.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;

public class MyXargs{
    private static List<String> commandArgs = new ArrayList<>();
    private static int num = -1; // used for -n # command
    private static String replaceIcommand = null; // -I {} command
    private static boolean rCommand = false; // -r command
    private static boolean tCommand = false; // -t command
    public static void main(String[] args){
        // check the number of arguments used and how to use MyXargs
        if(args.length < 1){
            errorUsage();
            return;
        }

        // go through args list to see which commands are used
        for(int i = 0; i < args.length; i++){
            switch (args[i]) {
                case "-n":
                    if(i + 1 < args.length){
                        try {
                            num = Integer.parseInt(args[++i]);
                        } catch (NumberFormatException e) {
                            errorUsage();
                            return;
                        }
                    }
                    break;
                case "-I":
                    if(i + 1 < args.length){
                        replaceIcommand = args[++i];
                    }
                case "-r":
                    rCommand = true;
                    break;
                case "-t":
                    tCommand = true;
                    break;
                default:
                    if(args[i].startsWith("-")){
                        errorUsage();
                        return;
                    }else{
                        commandArgs.add(args[i]);
                    }
                    break;
            }
        }

        List<String> inputs = readInput();
        if(inputs.isEmpty() && rCommand) return;
        // user tried to use -n and -I together
        if(num != -1 && replaceIcommand != null){
            System.out.println("-n and -I are mutually exclusive");
            return;
        }

        // check if user used the -I command
        if(replaceIcommand != null){
            withICommand(inputs);
        }else{
            withoutICommand(inputs);
        }
    
    }

    private static void errorUsage(){
        System.out.println("Usage: MyXargs [-n num] [-I replace] [-t] [-r] command");
    }

    // grab the output from the left side
    private static List<String> readInput(){
        List<String> inputs = new ArrayList<>();
        try(BufferedReader reader = new BufferedReader(new InputStreamReader(System.in))) {
            String line;
            while((line = reader.readLine()) != null){
                for(String word: line.trim().split("\\s+")){
                    inputs.add(sanitizeInput(word));
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return inputs;
    }

    // remove this characters from the input ;&|><*?()$
    private static String sanitizeInput(String input){
        return input.replaceAll("[;&|><*?()$]", "");
    }

    // this method will run if -I {} is provided
    private static void withICommand(List<String> inputs) {
    	for(List<String> batch : getBatches(inputs, num)) {
    		List<String> finalCommand = replacePlaceholders(commandArgs, batch);
    		if(tCommand) {
    			System.out.println(String.join(" ", finalCommand));
    		}
    		runCommand(finalCommand);
    	}
    }

    // this method will run if there is no -I {}
    private static void withoutICommand(List<String> inputs){
        for(List<String> batch : getBatches(inputs, num)){
            List<String> finalCommand = new ArrayList<>(commandArgs);
    		finalCommand.addAll(batch);
    		if(tCommand) {
    			System.out.println(String.join(" ",  finalCommand));
    		}
            runCommand(finalCommand);
        }
    }

    // For -I {} command
    // replace {} with the string
    private static List<String> replacePlaceholders(List<String> commandArguments, List<String> replacements) {
    	List<String> replacedCommand = new ArrayList<>();
    	for(String part : commandArguments) {
    		if(part.contains("{}")) {
    			for(String replacement : replacements) {
    				replacedCommand.add(part.replace("{}", replacement));
    			}
    		}
    		else {
    			replacedCommand.add(part);
    		}
    	}
    	return replacedCommand;
    }

    // combinne the inputs
    private static List<List<String>> getBatches(List<String> inputs, int num) {
    	List<List<String>> batches = new ArrayList<>();
    	if(num > 0) {
    		for(int i = 0; i < inputs.size(); i += num) {
    			batches.add(inputs.subList(i, Math.min(i + num, inputs.size())));
    		}
    	}
    	else {
    		batches.add(inputs);
    	}
    	return batches;
    }

    // run processBuilder to run all the command with arguments
    private static void runCommand(List<String> commandArgs){
        ProcessBuilder pb = new ProcessBuilder(commandArgs);
        pb.inheritIO();
        try {
            Process process = pb.start();
            process.waitFor();
        } catch (IOException | InterruptedException e) {
            e.printStackTrace();
        }
    }
}

File: csc139/myxargs.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#define MAX_INPUT_SIZE 1024
#define MAX_ARGS 100
#define MAX_COMMAND_LENGTH 256

typedef struct {
    char *args[MAX_ARGS];
    int count;
} ArgsList;

ArgsList commandArgs;
int num = -1; // used for -n # command
char *replaceIcommand = NULL; // -I {} command
bool rCommand = false; // -r command
bool tCommand = false; // -t command

void errorUsage() {
    printf("Usage: myxargs [-n num] [-I replace] [-t] [-r] command\n");
    exit(EXIT_FAILURE);
}

// sanitize special characters
char *sanitizeInput(char *input) {
    static char sanitized[MAX_INPUT_SIZE];
    int j = 0;
    for (int i = 0; input[i] != '\0'; i++) {
        if (strchr("; &|><*?()$", input[i]) == NULL) {
            sanitized[j++] = input[i];
        }
    }
    sanitized[j] = '\0';
    return sanitized;
}

// get input from stdin
void readInput(ArgsList *inputs) {
    char line[MAX_INPUT_SIZE];
    while (fgets(line, sizeof(line), stdin)) {
        char *token = strtok(line, " \t\n");
        while (token != NULL) {
            char *sanitized = sanitizeInput(token);
            inputs->args[inputs->count++] = strdup(sanitized);
            token = strtok(NULL, " \t\n");
        }
    }
}

// run command with the arguments
void runCommand(ArgsList *finalCommand) {
    if (tCommand) {
        for (int i = 0; i < finalCommand->count; i++) {
            printf("%s ", finalCommand->args[i]);
        }
        printf("\n");
    }

    char command[MAX_COMMAND_LENGTH];
    strcpy(command, finalCommand->args[0]);
    for (int i = 1; i < finalCommand->count; i++) {
        strcat(command, " ");
        strcat(command, finalCommand->args[i]);
    }

    system(command);
}

// used for replacing the {} in I command
void str_replace(char *str, const char *old, const char *new) {
    char buffer[MAX_COMMAND_LENGTH];
    char *p;

    if (!(p = strstr(str, old))) {
        return;
    }

    strncpy(buffer, str, p - str);
    buffer[p - str] = '\0';
    sprintf(buffer + (p - str), "%s%s", new, p + strlen(old));
    strcpy(str, buffer);
}

// -I command
void withICommand(ArgsList *inputs) {
    if (num == -1){
        num = 1;
    }
    for (int i = 0; i < inputs->count; i += num) {
        ArgsList batch;
        batch.count = 0;

        for (int j = i; j < i + num && j < inputs->count; j++) {
            batch.args[batch.count++] = inputs->args[j];
        }

        ArgsList finalCommand;
        finalCommand.count = 0;

        for (int k = 0; k < commandArgs.count; k++) {
            char *arg = commandArgs.args[k];
            if (strstr(arg, "{}")) {
                for (int l = 0; l < batch.count; l++) {
                    finalCommand.args[finalCommand.count] = malloc(MAX_COMMAND_LENGTH);
                    strcpy(finalCommand.args[finalCommand.count], arg);
                    str_replace(finalCommand.args[finalCommand.count], "{}", batch.args[l]);
                    finalCommand.count++;
                }
            } else {
                finalCommand.args[finalCommand.count++] = strdup(arg);
            }
        }

        runCommand(&finalCommand);
    }
}

// if -I command is not passed
void withoutICommand(ArgsList *inputs) {
    if (num == -1){
        num = 1;
    }
    for (int i = 0; i < inputs->count; i += num) {
        ArgsList batch;
        batch.count = 0;

        for (int j = i; j < i + num && j < inputs->count; j++) {
            batch.args[batch.count++] = inputs->args[j];
        }

        ArgsList finalCommand;
        finalCommand.count = 0;
        for (int k = 0; k < commandArgs.count; k++) {
            finalCommand.args[finalCommand.count++] = strdup(commandArgs.args[k]);
        }
        for (int l = 0; l < batch.count; l++) {
            finalCommand.args[finalCommand.count++] = strdup(batch.args[l]);
        }

        runCommand(&finalCommand);
    }
}

// Driver
int main(int argc, char *argv[]) {
    // check number of arguments and how to use program
    if (argc < 2) {
        errorUsage();
    }

    commandArgs.count = 0;

    // go through arg list
    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "-n") == 0) {
            if (i + 1 < argc) {
                num = atoi(argv[++i]);
            } else {
                errorUsage();
            }
        } else if (strcmp(argv[i], "-I") == 0) {
            if (i + 1 < argc) {
                replaceIcommand = argv[++i];
            } else {
                errorUsage();
            }
        } else if (strcmp(argv[i], "-r") == 0) {
            rCommand = true;
        } else if (strcmp(argv[i], "-t") == 0) {
            tCommand = true;
        } else {
            commandArgs.args[commandArgs.count++] = argv[i];
        }
    }

    ArgsList inputs;
    inputs.count = 0;
    readInput(&inputs);

    if (inputs.count == 0 && rCommand) {
        return 0;
    }

    if (num != -1 && replaceIcommand != NULL) {
        printf("-n and -I are mutually exclusive\n");
        return 1;
    }

    if (replaceIcommand != NULL) {
        withICommand(&inputs);
    } else {
        withoutICommand(&inputs);
    }

    return 0;
}
